game-core/src/test/java/games/strategy/triplea/util/MovableUnitsFilterTest.java

```diff
@@ -0,0 +1,115 @@
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
+import static org.hamcrest.core.IsNull.nullValue;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.PlayerId;
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitType;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.AbstractMoveDelegate;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.MockDelegateBridge;
+import games.strategy.triplea.xml.TestMapGameData;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Predicate;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.triplea.java.collections.CollectionUtils;
+
+public class MovableUnitsFilterTest {
+  final GameData gameData = TestMapGameData.REVISED.getGameData();
+  final PlayerId germans = germans(gameData);
+  final Territory germany = territory("Germany", gameData);
+  final Territory easternEurope = territory("Eastern Europe", gameData);
+  final Territory kareliaSsr = territory("Karelia S.S.R.", gameData);
+  final UnitType infantryType = infantry(gameData);
+  final UnitType armourType = armour(gameData);
+
+  public MovableUnitsFilterTest() throws Exception {}
+
+  private IDelegateBridge newDelegateBridge(final PlayerId player) {
+    return MockDelegateBridge.newInstance(gameData, player);
+  }
+
+  private MovableUnitsFilter.Result filterUnits(final Route route, final Collection<Unit> units) {
+    final IDelegateBridge bridge = newDelegateBridge(germans);
+    advanceToStep(bridge, "NonCombatMove");
+    final MovableUnitsFilter filter =
+        new MovableUnitsFilter(
+            gameData, germans, route, false, AbstractMoveDelegate.MoveType.DEFAULT, List.of());
+    return filter.filterUnitsThatCanMove(units, Map.of());
+  }
+
+  private Collection<Unit> germanyUnits(final Predicate<Unit> predicate) {
+    return CollectionUtils.getMatches(germany.getUnits(), predicate);
+  }
+
+  @Test
+  @DisplayName("moving infantry and tanks two territories should filter to just tanks")
+  void onlyTanksCanMoveTwo() {
+    final Route route = new Route(germany, easternEurope, kareliaSsr);
+    final Collection<Unit> units = germanyUnits(Matches.unitIsOfTypes(infantryType, armourType));
+    assertThat(units, hasSize(5));
+    final Collection<Unit> justTanks = germanyUnits(Matches.unitIsOfType(armourType));
+    assertThat(justTanks, hasSize(2));
+
+    final var result = filterUnits(route, units);
+    assertThat(result.getErrorMessage(), is(nullValue()));
+    assertThat(result.getWarningMessage(), is("Not all units have enough movement"));
```

Comment:
nit, brittle test code smell. If we change this string in production code, this test will fail unnecessarily. Recommend extracting the string to a constant, eg:
```
@VisibleForTesting
static String NOT_ALL_UNITS_CAN_MOVE = "Not all units have enough movement";
```

I wonder a bit too if maybe the string that is displayed should be a downstream concern. For example, if we were to internationalize the strings (not on our radar, but for the sake of argument), then it will be odd to have the dependency on the string value. Perhaps a cleaner API would return an enum result, then the consumer could map that result to a string. That would make this test read more like:

```
 assertThat(result.getValue(), is(SOME_UNITS_CAN_MOVE));
```

The other two enum values I would presume would be:
```
ALL_UNITS_CAN_MOVE
NO_UNITS_CAN_MOVE
```
This also avoids a good bit of nullable args. Nullable return types  are essentially a SRP violation code-smell, they represent values that have different meanings depending on the value. IE: if null, this variable means one thing (a flag), if not-null, it's a display string for the user. Having enum values would then imply non-null return type, and code that is closer to the front-end could decide about which display string to use.

Reply:
How much of the code do you want me to refactor in one PR? Changing this would require changing MoveValidator in a significant way. It has lots of other error messages it can give, such as "Not enough transports" and a bunch of other cases that should not be possible given how this called is called (e.g. "Can only move friendly units").

I'm trying to keep each PR to a reasonable size to make them easier to review and minimize risk, so I don't think this change should be done here.